widgets/src/main/java/org/pentaho/gwt/widgets/client/tabs/PentahoTab.java

```diff
@@ -109,13 +100,40 @@ public void onBrowserEvent( Event event ) {
     } else if ( event.getButton() == Event.BUTTON_RIGHT ) {
       onRightClick( event );
     } else if ( event.getButton() == Event.BUTTON_LEFT ) {
-      if ( event.getEventTarget().toString().toLowerCase().indexOf( "image" ) == -1 ) {
+      if ( KeyCodes.KEY_LEFT == (char) event.getKeyCode() ) {
+        moveFocusToPreviousTab();
+      } else if ( KeyCodes.KEY_RIGHT == (char) event.getKeyCode() ) {
+        moveFocusToNextTab();
+      } else if ( event.getEventTarget().toString().toLowerCase().indexOf( "image" ) == -1 ) {
         fireTabSelected();
       }
+
     }
     super.onBrowserEvent( event );
   }
 
+  private void moveFocusToNextTab() {
+    FlowPanel tabBar = getTabPanel().getTabBar();
+    for ( int i = 0; i < tabBar.getWidgetCount(); i++ ) {
+      PentahoTab tab = (PentahoTab) tabBar.getWidget(i);
+      if ( getTabPanel().selectedTab == tab && i < tabBar.getWidgetCount() ) {
+        tabPanel.selectTab( (PentahoTab) tabBar.getWidget( i + 1 ) );
+        break;
+      }
+    }
+  }
+
+  private void moveFocusToPreviousTab() {
+    FlowPanel tabBar = getTabPanel().getTabBar();
+    for ( int i = 0; i < tabBar.getWidgetCount(); i++ ) {
+      PentahoTab tab = (PentahoTab) tabBar.getWidget(i);
+      if ( getTabPanel().selectedTab == tab && i > 0 ) {
+        tabPanel.selectTab( (PentahoTab) tabBar.getWidget( i - 1 ) );
+        break;
+      }
+    }
+  }
```

Comment:
All this logic should be placed in the `PentahoTabPanel` class — it belongs more to the context in which the button/tab is used. `PentahoTabPanel` should listen to the relevant events.
If needed, define a public method such as `addKeyDownHandler`.

Reply:
Hi @dcleao. from the comments added for PentahoTab class and PentahoTabPanel class whatever changes you are suggesting are not feasible. PentahoTabPanel is the entire row of tabs and PentahoTab is each tab inside it.

we have to use listeners, tabindex, focus on each PentahoTab not the panel. And focus as you mentioned an option is a must not an option. you can refer here https://www.w3.org/WAI/ARIA/apg/example-index/tabs/tabs-automatic.html
and https://www.w3.org/WAI/ARIA/apg/example-index/tabs/js/tabs-automatic.js

tab.addEventListener('keydown', this.onKeydown.bind(this));

the EventListener is added to the Tab right ?

tab.focus();

the focus is set on Tab right ?

I am not sure what are we going to achieve out of implementing Focusable ? as the code is already working and unit tested as per the standard. Implementing Focusable is an option, but is it must ? really ? needed ???

Reply:
We have talked, in the meantime, and I believe that you now agree that your response does not apply to what I was referring to.

In any case, for the record. What I was saying is independent of having the manual or automatic selection pattern, which concerns what happens when you tab into a tab button, whether the corresponding tab is activated, selected, becomes the current one. My comment concerns what happens in the other direction: whether selecting a tab automatically places the focus into it. You may wish to (programatically) make a tab become current, without placing the focus in its tab button (and stealing the focus from wherever it is).